src/box/applier.cc

```diff
@@ -499,6 +499,8 @@ applier_wait_snapshot(struct applier *applier)
 		coio_read_xrow(io, ibuf, &row);
 	}
 
+	applier_set_state(applier, APPLIER_CONTINUE_SNAPSHOT);
```

Comment:
The name is really confusing IMO. 'Continue' means it was stopped. But it wasn't. Intuitively, FETCH_SNAPSHOT is the good name, but it is occupied. I would suggest to reshuffle the names a bit:
- `APPLIER_FETCH_SNAPSHOT` -> `APPLIER_WAIT_SNAPSHOT`;
- `APPLIER_CONTINUE_SNAPSHOT` -> `APPLIER_FETCH_SNAPSHOT`.

The first rename should be done in a separate patch. The only worrying part is that there might be some tests using `fetch_snapshot` or whatever it is stringified to. They could be fixed. But I would also recommend to check if the state `fetch_snapshot` is documented anywhere on our web-site. If it is, then we need to think more about whether it can be renamed.

Reply:
Please, check out my proposal
https://github.com/tarantool/tarantool/pull/7183#discussion_r907215400

Reply:
I can't see any verification that the 'continue' is performed using the same snapshot. Between the unfinished fetch and the continuation the snapshot can be done on the relay side. This new snapshot can contain some new data with PK our local applier consider as 'done'.

Reply:
I don't understand. No data is applied before `APPLIER_CONTINUE_SNAPSHOT` is set, and no retries are possible after `APPLIER_CONTINUE_SNAPSHOT`. So the applier will either receive a whole snapshot, or receive nothing at all (simply some metadata, which will be overwritten after a retry)

Reply:
This patch doesn't help when a snapshot receipt is interrupted in the middle due to some error. It's not easy for the reason you mention (about new data for already processed PK) and also because vinyl creates files which can't be easily overwritten or amended.

So this patch only lets replica retry joining when an error occurs before master starts sending the snapshot

Reply:
Done